### DISCUSSION TITLE
Migrating from 1.x to 4.0, from DateTime to Temporal

### DISCUSSION THREAD

**Discussion by alxtlst** [ical4j/ical4j#664] [Q&A] [answered]
Hello,
I'm working on an application that have previously used an older version of ical4j, and I'm migrating it to the latest version. The relevant part of the application parses .ics files to an internal data format, which stores start and end of an event as java.time.Instant. In the new version DtStart and DtEnd objects contain Temporal values, but I can't find any way to programmatically determine what exact implementation of Temporal is contained within a given DtStart/DtEnd instance parsed from a file, so I can't really do anything useful with these objects.

Could someone give some pointers how to handle net.fortuna.ical4j.model.property.DateProperty subclasses and how to work with their values?

**Comment by benfortuna** [comment, marked as answer]
Unfortunately the usage documentation for v4 is still quite lacking, but you can usually get a sense of how to use it from the unit tests. That said, you can assume one of three Temporal implementations to be used in DtStart/DtEnd when parsed from a file:

1. LocalDate/LocalDateTime - when property is a date or uses a floating date-time format
2. ZonedDateTime - when property includes a TZID parameter
3. Instant - when property is specified in UTC (suffix is Z)

There is limited support for identifying the type, such as `DateProperty.isUtc()`, but the easiest way to determine the type is via an 'instanceof' check.

**Reply by alxtlst** [reply]
Thank you very much!

Just to make sure I understood you correctly: 
Temporal instance inside of a DateProperty instance always belongs to one of the following classes: LocalDate, LocalDateTime, ZonedDateTime or Instant, so I may safely assume that four "if (x instanceof Y)" will be able to handle any value parsed from a file, right?

**Reply by benfortuna** [reply]
Yes correct. All date strings are parsed via [CalendarDateFormat](https://github.com/ical4j/ical4j/blob/feature/jsr-310-localdate/src/main/java/net/fortuna/ical4j/model/CalendarDateFormat.java) which results in one of those four types.
